src/ast/passes/semantic_analyser.cpp

```diff
@@ -2824,35 +2834,59 @@ void SemanticAnalyser::visit(ExprStatement &expr)
   Visit(expr.expr);
 }
 
+static const std::unordered_map<Type, std::string_view> AGGREGATE_USAGES{
+  { Type::count_t, "count()" },
+  { Type::sum_t, "sum(retval)" },
+  { Type::min_t, "min(retval)" },
+  { Type::max_t, "max(retval)" },
+  { Type::avg_t, "avg(retval)" },
+  { Type::hist_t, "hist(retval)" },
+  { Type::lhist_t, "lhist(rand %10, 0, 10, 1)" },
+  { Type::stats_t, "stats(arg2)" },
+};
+
 void SemanticAnalyser::visit(AssignMapStatement &assignment)
 {
   Visit(assignment.map);
   Visit(assignment.expr);
 
-  auto type = assignment.expr->type;
-
-  if (!IsValidAssignment(type, assignment.expr->is_map)) {
-    auto hint = "";
-    if (type.IsHistTy()) {
-      hint = "hist(retval)";
-    } else if (type.IsLhistTy()) {
-      hint = "lhist(rand %10, 0, 10, 1)";
-    } else if (type.IsStatsTy()) {
-      hint = "stats(arg2)";
-    } else {
+  const auto *map_type = get_map_type(*assignment.map);
+  const bool is_map_declared = map_type != nullptr;
```

Comment:
maybe this is better or `map_has_type`?
```suggestion
  const bool is_map_assigned = map_type != nullptr;
```